csi: mount host's /etc/selinux in node plugins

humblec opened the pull request:
This commit introduces a new configuration option for
    ceph csi driver to enable hostpath mounting of /etc/selinux
    directory from the cluster node where csi plugin pods are
    running, which inturn help the csi driver to specify
    selinux-related mount options like context.
    
    Ref# https://github.com/ceph/ceph-csi/issues/2295
    
    The default value for this configuration is true and if cluster
    nodes are running without selinux enabled, an admin can deploy
    csi pods by specifying this option to `false` which skip the
    host path mounting for the csi pods.


Signed-off-by: Humble Chirammal <hchiramm@redhat.com>



**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

humblec commented:
@Madhu-1 ptal.. thanks

humblec commented on cluster/examples/kubernetes/ceph/csi/template/cephfs/csi-cephfsplugin.yaml:
```diff
@@ -105,6 +105,9 @@ spec:
               mountPath: /tmp/csi/keys
             - name: host-run-mount
               mountPath: /run/mount
+            - name: etc-selinux
```
@travisn that make sense,  anything specifically in your mind to make this optional mount?

humblec replied:
yea, so in the latesr verison of this patch a condition is introduced which read from the csi setting.

leseb commented:
@humblec any plan to resume this?

humblec commented:
> @humblec any plan to resume this?

@leseb @travisn  was planning to get these changes along with next ceph csi release template (3.4.1) changes.. so kept in this state.. sorry for missing to update the plan here.

leseb commented:
> > @humblec any plan to resume this?
> 
> @leseb @travisn we were planning to get these changes along with next ceph csi release template (3.4.1) changes.. so kept this PR with this state.. sorry for missing to update the plan here.

No worries, what should we do with this PR? Thanks!

humblec commented:
> > > @humblec any plan to resume this?
> > 
> > 
> > @leseb @travisn we were planning to get these changes along with next ceph csi release template (3.4.1) changes.. so kept this PR with this state.. sorry for missing to update the plan here.
> 
> No worries, what should we do with this PR? Thanks!

will update this PR addressing the comments and also an extra commit with the release template changes once the release is out  .. hope it works

humblec commented:
> @humblec any plans to pursue this? Thanks

Let me share some details on what was happening on this meanwhile.. That said, this change is introduced here to follow up a change in ceph csi https://github.com/ceph/ceph-csi/issues/2295 which address an issue with long pod start delay if the pod has a big volume with many files in it. 
Considering this bind mount and this context population could help in OCP platform this was introduced as soon as the ceph csi PR merged.  On later experiments it has seen that, the selinux get enabled by default in OCP platform based on the container build. In that case, the fix we were targetting will work on OCP,  ie even without
bind mounting this share and contexts are populated as expected. When it comes to the upstream, this depend on where csi pods are getting spawned and if we see an selinux enabled system we have to bind mount, otherwise skip the mount. Indeed that was the suggestion here ( from @travisn ) and a good to have the fix. 
EOD, if systems are without selinx, the subjected fixed security context any way does not help , rather users will not be hitting pod startup delay due to 'selinux'

leseb commented on pkg/operator/ceph/csi/template/cephfs/csi-cephfsplugin.yaml:
```diff
@@ -105,6 +105,11 @@ spec:
               mountPath: /tmp/csi/keys
             - name: host-run-mount
               mountPath: /run/mount
+            { { if .EnablePluginSelinuxHostMount } }
```
```suggestion
            {{ if .EnablePluginSelinuxHostMount }}
```

humblec replied:
Done.. thanks !

leseb commented on pkg/operator/ceph/csi/template/cephfs/csi-cephfsplugin.yaml:
```diff
@@ -105,6 +105,11 @@ spec:
+            -  name: etc-selinux
+               mountPath: /etc/selinux
+               readOnly: true
+            { { end } }
```
```suggestion
            {{ end }}
```

leseb commented on pkg/operator/ceph/csi/template/rbd/csi-rbdplugin.yaml:
```diff
@@ -106,6 +106,11 @@ spec:
               mountPath: /tmp/csi/keys
             - name: host-run-mount
               mountPath: /run/mount
+            { { if .EnablePluginSelinuxHostMount } }
```
```suggestion
            {{ if .EnablePluginSelinuxHostMount }}
```

leseb commented on pkg/operator/ceph/csi/template/rbd/csi-rbdplugin.yaml:
```diff
@@ -106,6 +106,11 @@ spec:
+            -  name: etc-selinux
+               mountPath: /etc/selinux
+               readOnly: true
+            { { end } }
```
```suggestion
            {{ end }}
```

Madhu-1 commented on cluster/charts/rook-ceph/values.yaml:
```diff
@@ -66,6 +66,8 @@ csi:
   # set to false to disable deployment of snapshotter container in RBD provisioner pod.
   enableRBDSnapshotter: true
+  # set to false if the selinux is not enabled or unavailable in cluster nodes.
+  enablePluginSelinuxHostMount : true
```
This need to be documented  in helm doc

Madhu-1 commented on cluster/examples/kubernetes/ceph/operator-openshift.yaml:
```diff
@@ -154,6 +154,10 @@ data:
   # (Optional) Allow starting unsupported ceph-csi image
   ROOK_CSI_ALLOW_UNSUPPORTED_VERSION: "false"
+
+  # (Optional) control the host mount of /etc/selinux for csi plugin pods.
```
don't we need to document when this needs to be enabled?

humblec replied:
we are defaulting to true here.

Madhu-1 commented on pkg/operator/ceph/csi/spec.go:
```diff
@@ -463,6 +468,7 @@ func (r *ReconcileCSI) startDrivers(ver *version.Info, ownerInfo *k8sutil.OwnerI
 		// get CephFS plugin tolerations and node affinity, defaults to common tolerations and node affinity if not specified
 		cephFSPluginTolerations := getToleration(r.opConfig.Parameters, cephFSPluginTolerationsEnv, pluginTolerations)
 		cephFSPluginNodeAffinity := getNodeAffinity(r.opConfig.Parameters, cephFSPluginNodeAffinityEnv, pluginNodeAffinity)
+
```
unnecessary change not needed?

Madhu-1 commented on pkg/operator/ceph/csi/template/cephfs/csi-cephfsplugin.yaml:
```diff
@@ -166,3 +171,6 @@ spec:
         - name: host-run-mount
           hostPath:
             path: /run/mount
+        - name: etc-selinux
```
this should also have if check

Madhu-1 commented on pkg/operator/ceph/csi/template/rbd/csi-rbdplugin.yaml:
```diff
@@ -176,3 +181,6 @@ spec:
         - name: host-run-mount
           hostPath:
             path: /run/mount
+        - name: etc-selinux
```
same here

Madhu-1 commented on pkg/operator/ceph/csi/template/rbd/csi-rbdplugin.yaml:
```diff
@@ -176,3 +181,6 @@ spec:
             path: /run/mount
+        - name: etc-selinux
+          hostPath:
+            path: /etc/selinux
```
will this path be present on all the distro and this path need to be configurable?

humblec replied:
I didnt get the question @Madhu-1 , can you please elaborate?

Madhu-1 replied:
Is the `/etc/selinux` path present on all the Linux distributions? if not deployment will fail?

humblec replied:
if selinux available system , it will be present. thus guaranteed to be working in platforms like openshift. For kube, we are defaulting to `false` and user can enable, if they are running selinux enabled cluster and if they want . does that answer @Madhu-1 ?

Madhu-1 replied:
my question is is it always `/etc/selinux`? i see from some link its at `/etc/sysconfig/selinux` if it's not constant this need to be a configurable parameter for the user

humblec replied:
@Madhu-1 if you can share I can check the link, but to my knowledge its a symlink and we dont have to worry about it.

Madhu-1 replied:
here are the links I found about it https://www.strongdm.com/docs/architecture/deployment/selinux https://www.omnisecu.com/gnu-linux/redhat-certified-engineer-rhce/security-enhanced-linux-selinux-configuration-file.php

humblec replied:
as in the link https://www.omnisecu.com/gnu-linux/redhat-certified-engineer-rhce/security-enhanced-linux-selinux-configuration-file.php ```The SELinux configuration file is “/etc/sysconfig/selinux”, which is a symbolic link to “/etc/selinux/conf” file.``` , so should be fine.

humblec commented on cluster/charts/rook-ceph/values.yaml:
```diff
@@ -66,6 +66,8 @@ csi:
   # set to false to disable deployment of snapshotter container in RBD provisioner pod.
   enableRBDSnapshotter: true
+  # set to false if the selinux is not enabled or unavailable in cluster nodes.
+  enablePluginSelinuxHostMount : true
```
@travisn even in kubernetes, isnt it a good idea to default to `true` ?

humblec replied:
may be not, and good to be on false side here.

humblec replied:
agreed! patch carry the same. thanks.

humblec replied:
addressed, thanks.

humblec commented on cluster/examples/kubernetes/ceph/operator-openshift.yaml:
```diff
@@ -154,6 +154,10 @@ data:
   ROOK_CSI_ALLOW_UNSUPPORTED_VERSION: "false"
+
+  # (Optional) control the host mount of /etc/selinux for csi plugin pods.
+  # CSI_PLUGIN_ENABLE_SELINUX_HOST_MOUNT: "false"
```
yes, it is good to be enabled by default.

Madhu-1 commented on pkg/operator/ceph/csi/template/rbd/csi-rbdplugin.yaml:
```diff
@@ -106,6 +106,11 @@ spec:
               mountPath: /tmp/csi/keys
             - name: host-run-mount
               mountPath: /run/mount
+            {{ if .EnablePluginSelinuxHostMount }}
```
@humblec https://github.com/ceph/ceph-csi/issues/2295 is reported only for cephfs. Do we need changes in rbd alss?

humblec replied:
Enabling selinux relabelling is equally applicable for rbd too. so I have added it for rbd also.  @Madhu-1

humblec commented:
The test failures looks spurious,

humblec commented:
@travisn @leseb @Madhu-1 any further comments here ?

humblec commented on pkg/operator/ceph/csi/spec.go:
```diff
@@ -316,6 +317,10 @@ func (r *ReconcileCSI) startDrivers(ver *version.Info, ownerInfo *k8sutil.OwnerI
 		tp.RBDPluginUpdateStrategy = rollingUpdate
 	}
 
+	if strings.EqualFold(k8sutil.GetValue(r.opConfig.Parameters, "CSI_PLUGIN_ENABLE_SELINUX_HOST_MOUNT", "true"), "false") {
```
correct .

humblec commented on pkg/operator/ceph/csi/spec.go:
```diff
@@ -316,6 +317,10 @@ func (r *ReconcileCSI) startDrivers(ver *version.Info, ownerInfo *k8sutil.OwnerI
 	}
 
+	if strings.EqualFold(k8sutil.GetValue(r.opConfig.Parameters, "CSI_PLUGIN_ENABLE_SELINUX_HOST_MOUNT", "true"), "false") {
+		tp.EnableSelinuxHostMount = false
```
done.

humblec commented on pkg/operator/ceph/csi/template/rbd/csi-rbdplugin.yaml:
```diff
@@ -106,6 +106,11 @@ spec:
             - name: host-run-mount
               mountPath: /run/mount
+            {{ if .EnablePluginSelinuxHostMount }}
+            -  name: etc-selinux
```
Thanks @travisn , thats been corrected. looks like tests are still failing.. so need to check on it..

mykaul commented:
Is there an implicit assumption here that the selinux policy on the host and in the containers is the same? What if they are (for example), running different RHEL versions? Did we consult the selinux team on this?

humblec commented:
> Is there an implicit assumption here that the selinux policy on the host and in the containers is the same? What if they are (for example), running different RHEL versions? 

Good question @mykaul , actually that was  the scenario where we have started with. That said, while we have  different `libmount` ( util-lnux)  version to be precise, thus similar to a situation of  different rhel version in container and host scenario we have seen that, the selinux contexts were not getting respected, thus not able to make the context population work properly.  So the host mount fix was introduced.  

Apart from above, have gone through the selinux code and from the selinux code reading what have been identified is that, this mount is actually inspected just for selinux enablement part. (via the `is_selinux_enabled` and `selinux_mnt` ( set by `init_selinuxmnt`->`verify_selinuxmnt `which make the decision based on `SELINUXMNT and OLDSELINUXMNT` variables [1]), thats where I was thinking this is safe to bind mount and also being its `readonly` mount , it should be fine. The other decision point which selinux make use of in the code  is based on `/sys`  which any way our node plugin pods are mounting from the host almost since start of csi driver. To summarize,  even if host and containers builds are different, it looks like we are good. 

>Did we consult the selinux team on this?

havent consulted selinux team specifically on this and only bit of code reading in specific areas and few reports from upstream users that, things are working with this bind mount and no complaints so far we have heard. However I can also check with selinux team and double confirm about this approach. 

[1] https://github.com/SELinuxProject/selinux/blob/d155b410d4bbc90d28f361b966f0429598da8188/libselinux/src/enabled.c

leseb commented:
@humblec isn't `/sys/fs/selinux` sufficient for `is_selinux_enabled` to return true? Why is `/etc/selinux` adding to this detection? Also, I left a comment for clarification here https://github.com/ceph/ceph-csi/issues/2295#issuecomment-980048956, thanks!

humblec commented:
> @humblec isn't `/sys/fs/selinux` sufficient for `is_selinux_enabled` to return true? Why is `/etc/selinux` adding to this detection? Also, I left a comment for clarification here [ceph/ceph-csi#2295 (comment)](https://github.com/ceph/ceph-csi/issues/2295#issuecomment-980048956), thanks!

selinux also check for the config in its internal code path, so this helps.

humblec commented:
> The changes look good, but the CI is failing, rbd pvs aren't being bound. I don't see a pod description of why the test pod is stuck pending, so it may be easier to test your changes locally than debugging the CI. We should also verify the changes on OpenShift with this setting enabled since we don't have that env in the CI.

@travisn  So all tests are passing and also verified this in a kubernetes setup with the default value and verified PVCs are bound and working properly.  The OCP validation with flipping the value to `true` has not been done yet though.

humblec commented:
> To resolve the merge conflicts, note the following changes from #8740:
> 
>     * `cluster/examples/kubernetes/ceph/` moved to `deploy/examples`
> 
>     * `cluster/charts/` moved to `deploy/charts`

Rebased to master.. Thanks , ptal..
